### DISCUSSION TITLE
How to use dynamic app layout with typescript

### DISCUSSION THREAD

**Discussion by callumbooth** [vercel/next.js#18502] [Help]
Based on the [with-dynamic-app-layout](https://github.com/vercel/next.js/tree/canary/examples/with-dynamic-app-layout) how best should this be used with typescript?

Without some declaration merging you will get ts compile errors saying something like `Layout doesn't exist on...`

It looks the `Component` prop passed into the _app component is using the `NextComponentType` and we can use declaration merging to add the Layout property onto the FunctionComponent and ComponentClass types something like this:

```ts
declare module 'react' {
  interface FunctionComponent {
    Layout?: ReactNode;
  }
  
  interface ComponentClass {
    ReactNode?: ReactNode;
  }
}
```

However this will add `ReactNode` to all react components when really we only want allow it on components that are typed with `NextPage` or `NextComponentType`.

So what is the best way to use dynamic app layouts with typescript?

**Comment by MathiasKandelborg** [comment]
You might find this section useful: https://nextjs.org/docs/basic-features/typescript#custom-app

Here's how I typed `_app` with it being a classic function: https://github.com/MathiasKandelborg/TRUE-framework/blob/master/pages/_app.tsx#L15-L25

**Comment by bryanveloso** [comment]
I ran into this same problem today and happened upon a starter (https://github.com/ippo012/nextjs-starter/blob/master/src/types/next.d.ts) that used this method to extend:

```
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { NextComponentType, NextPageContext, NextLayoutComponentType } from 'next';
import type { AppProps } from 'next/app';

declare module 'next' {
  type NextLayoutComponentType<P = {}> = NextComponentType<NextPageContext, any, P> & {
    getLayout?: (page: ReactNode) => ReactNode;
  };

  type NextLayoutPage<P = {}, IP = P> = NextComponentType<NextPageContext, IP, P> & {
    getLayout: (page: ReactNode) => ReactNode;
  };
}

declare module 'next/app' {
  type AppLayoutProps<P = {}> = AppProps & {
    Component: NextLayoutComponentType;
  };
}
```

Granted I'm nowhere near what you'd call proficient at TypeScript, but it got rid of the warning for me. I am using `getLayout()` like the starter, but I'm sure it can be tuned to Next's example. 😅
